### ISSUE TITLE
Error executing process > 'POREC:makeReport (1)'

### ISSUE THREAD

**Issue by hgingras** [epi2me-labs/wf-pore-c#18]
### What happened?

I am trying to configure the environment to run wf-pore-c v0.0.8 with the test-data. 

There is something wrong with the makeReport step.

Here is my script:
nextflow run main.nf -c nextflow.config  --fastq test_data/porec_test.concatemers.fastq  --ref test_data/porec_test.fasta  --phased_vcf test_data/porec_test.phased_variants.vcf.gz  --cutter NlaIII  --vcf test_data/porec_test.phased_variants.vcf.gz --pairs  --chromunity --paired_end --filter_pairs --paired_end_minimum_distance 100  --paired_end_maximum_distance 200  --out_dir outputPoreC-test  -profile local

I have installed fastcat version1 and  mosdepth-0.3.4. 
Modules that I load:
module load StdEnv/2020 python/3.10.2 gcc/9.3.0 arrow/11.0.0 nextflow/22.10.8 minimap2/2.24 samtools/1.17

Here is the content of my virtual environment:

anyio                              3.6.2+computecanada
arff                               0.9+computecanada
argon2_cffi                        21.3.0+computecanada
argon2_cffi_bindings               21.2.0+computecanada
asciitree                          0.3.3+computecanada
asttokens                          2.2.1+computecanada
async_generator                    1.10+computecanada
async_timeout                      4.0.2+computecanada
attrs                              22.2.0+computecanada
backcall                           0.2.0+computecanada
backports-abc                      0.5+computecanada
backports.shutil_get_terminal_size 1.0.0+computecanada
bcrypt                             4.0.1+computecanada
beautifulsoup4                     4.11.2+computecanada
bioframe                           0.3.3+computecanada
biopython                          1.79+computecanada
bitstring                          4.0.1+computecanada
bleach                             6.0.0+computecanada
certifi                            2022.12.7+computecanada
cffi                               1.15.1+computecanada
chardet                            5.1.0+computecanada
charset_normalizer                 3.1.0+computecanada
click                              8.1.4+computecanada
comm                               0.1.2+computecanada
contourpy                          1.0.7+computecanada
cooler                             0.8.11+computecanada
cryptography                       39.0.1+computecanada
cycler                             0.11.0+computecanada
Cython                             0.29.33+computecanada
cytoolz                            0.10.1
deap                               1.3.3+computecanada
debugpy                            1.6.6+computecanada
decorator                          5.1.1+computecanada
defusedxml                         0.7.1+computecanada
dill                               0.3.6+computecanada
dnspython                          2.3.0+computecanada
dominate                           2.8.0
ecdsa                              0.18.0+computecanada
entrypoints                        0.4+computecanada
executing                          1.2.0+computecanada
ezcharts                           0.5.4
fastjsonschema                     2.16.2+computecanada
fonttools                          4.40.0+computecanada
funcsigs                           1.0.2+computecanada
h5py                               3.8.0+computecanada
idna                               3.4+computecanada
importlib_metadata                 5.2.0+computecanada
importlib_resources                5.12.0+computecanada
ipykernel                          6.21.2+computecanada
ipython                            8.10.0+computecanada
ipython_genutils                   0.2.0+computecanada
ipywidgets                         8.0.4+computecanada
jedi                               0.18.2+computecanada
Jinja2                             3.1.2+computecanada
jsonschema                         4.17.3+computecanada
jupyter_client                     8.0.0
jupyter_core                       5.2.0+computecanada
jupyter_events                     0.6.3+computecanada
jupyter_server                     2.3.0+computecanada
jupyter_server_terminals           0.4.4+computecanada
jupyterlab_pygments                0.2.2+computecanada
jupyterlab_widgets                 3.0.5+computecanada
kiwisolver                         1.4.4+computecanada
libsass                            0.22.0
lockfile                           0.12.2+computecanada
MarkupSafe                         2.1.2+computecanada
matplotlib                         3.7.0+computecanada
matplotlib_inline                  0.1.6+computecanada
mistune                            2.0.5+computecanada
mock                               5.0.1+computecanada
mpmath                             1.2.1+computecanada
multiprocess                       0.70.14+computecanada
nbclassic                          0.5.2+computecanada
nbclient                           0.7.2+computecanada
nbconvert                          7.2.9+computecanada
nbformat                           5.7.3+computecanada
nest_asyncio                       1.5.6+computecanada
netaddr                            0.8.0+computecanada
netifaces                          0.11.0+computecanada
nose                               1.3.7+computecanada
notebook                           6.5.2+computecanada
notebook_shim                      0.2.2+computecanada
numpy                              1.23.2+computecanada
packaging                          23.1+computecanada
pairtools                          1.0.1+computecanada
pandas                             1.5.3+computecanada
pandocfilters                      1.5.0+computecanada
paramiko                           3.0.0+computecanada
parso                              0.8.3+computecanada
path                               16.6.0+computecanada
path.py                            12.5.0+computecanada
pathlib2                           2.3.7+computecanada
paycheck                           1.0.2+computecanada
pbr                                5.11.1+computecanada
pexpect                            4.8.0+computecanada
pickleshare                        0.7.5+computecanada
Pillow                             9.5.0+computecanada
pip                                23.1.2
pkgutil_resolve_name               1.3.10+computecanada
platformdirs                       3.2.0+computecanada
pore-c-py                          2.0.5
prometheus_client                  0.16.0+computecanada
prompt_toolkit                     3.0.37+computecanada
psutil                             5.9.4+computecanada
ptyprocess                         0.7.0+computecanada
pure_eval                          0.2.2+computecanada
pyarrow                            11.0.0
pycountry                          22.3.5+computecanada
pycparser                          2.21+computecanada
pydantic                           1.10.11
pyfaidx                            0.7.2.1+computecanada
Pygments                           2.14.0+computecanada
PyNaCl                             1.5.0+computecanada
pypairix                           0.3.7+computecanada
pyparsing                          3.0.9+computecanada
pyrsistent                         0.19.3+computecanada
pysam                              0.20.0+computecanada
python-dateutil                    2.8.2+computecanada
python_json_logger                 2.0.7+computecanada
pytz                               2023.3+computecanada
PyYAML                             6.0+computecanada
pyzmq                              25.0.0+computecanada
redis                              4.6.0
requests                           2.28.2+computecanada
rfc3339_validator                  0.1.4+computecanada
rfc3986_validator                  0.1.1+computecanada
scipy                              1.10.1+computecanada
seaborn                            0.12.2+computecanada
Send2Trash                         1.8.0+computecanada
setuptools                         67.7.2
sigfig                             1.3.2
simplegeneric                      0.8.1+computecanada
simplejson                         3.19.1+computecanada
singledispatch                     4.0.0+computecanada
six                                1.16.0+computecanada
sniffio                            1.3.0+computecanada
sortedcontainers                   2.4.0+computecanada
soupsieve                          2.4+computecanada
stack_data                         0.6.2+computecanada
sympy                              1.11.1+computecanada
terminado                          0.17.1+computecanada
testpath                           0.6.0+computecanada
tinycss2                           1.2.1+computecanada
toolz                              0.12.0+computecanada
tornado                            6.2+computecanada
traitlets                          5.9.0+computecanada
typing_extensions                  4.5.0+computecanada
tzdata                             2023.3+computecanada
urllib3                            1.26.15+computecanada
wcwidth                            0.2.6+computecanada
webencodings                       0.5.1+computecanada
websocket_client                   1.5.1+computecanada
wheel                              0.38.4+computecanada
widgetsnbextension                 4.0.5+computecanada
zipp                               3.14.0+computecanada

### Operating System

ubuntu 18.04

### Workflow Execution

Command line

### Workflow Execution - CLI Execution Profile

None

### Workflow Version

epi2me-labs/wf-pore-c v0.0.8

### Relevant log output

```shell
executor >  local (8)
[72/e026be] process > POREC:fastcat (1)           [100%] 1 of 1 ✔
[bd/bf6ba4] process > POREC:index_ubam (1)           [100%] 1 of 1 ✔
[41/15b24c] process > POREC:prepare_genome:index_ref_mmi (1)[100%] 1 of 1 ✔
[-        ] process > POREC:digest_align_annotate (1)        -
[-        ] process > POREC:haplotag_alignments              -
[-        ] process > POREC:merge_coordsorted_bams           -
[-        ] process > POREC:merge_namesorted_bams            -
[7d/9c2577] process > POREC:create_restriction_bed (1) [100%] 1 of 1 ✔
[-        ] process > POREC:to_pairs_file                    -
[-        ] process > POREC:merge_pairs                      -
[-        ] process > POREC:merge_pairs_stats                -
[-        ] process > POREC:pair_stats_report                -
[-        ] process > POREC:merge_parquets_to_dataset        -
[-        ] process > POREC:merge_paired_end_bams            -
[d3/f9aae2] process > POREC:getVersions       [100%] 1 of 1 ✔
[8c/4df3a0] process > POREC:getParams             [100%] 1 of 1 ✔
[93/9cf9bd] process > POREC:makeReport (1)   [100%] 1 of 1, failed: 1 ✘
WARN: Access to undefined parameter `summary` -- Initialise it to a default value eg. `params.summary = some_value`
Error executing process > 'POREC:makeReport (1)'

Caused by:
  Process `POREC:makeReport (1)` terminated with an error exit status (1)

Command executed:

  source  /home/helene/scratch/Ticket/wf-pore-C/pore-c.sh
      echo '[
      {
          "barcode": null,
          "type": "test_sample",
          "alias": "porec_test"
      }
  ]' > metadata.json
      workflow-glue report wf-pore-c-report.html         --metadata metadata.json         --stats per_read_stats.tsv         --versions versions         --params params.json

Command exit status:
  1

Command output:
  (empty)

Command error:
  
  Lmod a automatiquement remplacé "intel/2020.1.217" par "gcc/9.3.0"
  
  
  Lmod a automatiquement remplacé "intel/2020.1.217" par "gcc/9.3.0"
  
  /lustre04/scratch/helene/Ticket/wf-pore-C/ENV_pore-c/lib/python3.10/site-packages/jupyter_client/connect.py:20: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write
  [15:04:17 - workflow_glue] Starting entrypoint.
  Traceback (most recent call last):
    File "/lustre04/scratch/helene/Ticket/wf-pore-C/wf-pore-c/bin/workflow-glue", line 7, in <module>
      cli()
    File "/lustre04/scratch/helene/Ticket/wf-pore-C/wf-pore-c/bin/workflow_glue/__init__.py", line 62, in cli
      args.func(args)
    File "/lustre04/scratch/helene/Ticket/wf-pore-C/wf-pore-c/bin/workflow_glue/report.py", line 31, in main
      fastcat.SeqSummary(args.stats)
    File "/lustre04/scratch/helene/Ticket/wf-pore-C/ENV_pore-c/lib/python3.10/site-packages/ezcharts/components/fastcat.py", line 47, in __init__
      df_all = load_stats(seq_summary, format='fastcat')
    File "/lustre04/scratch/helene/Ticket/wf-pore-C/ENV_pore-c/lib/python3.10/site-packages/ezcharts/components/fastcat.py", line 333, in load_stats
      if os.path.isdir(stat):
    File "/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx512/Core/python/3.10.2/lib/python3.10/genericpath.py", line 42, in isdir
      st = os.stat(s)
  TypeError: stat: path should be string, bytes, os.PathLike or integer, not list

Work dir:
  /lustre04/scratch/helene/Ticket/wf-pore-C/wf-pore-c/work/93/9cf9bdeb2383d3875f312d7fece700

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`
```

**Comment by sarahjeeeze** [comment]
Hi, The workflow is untested with python 3.10 our docker container for this workflow has python pinned to 3.8.*. You could use the docker or singularity containers? Or try an environment with python 3.8.16

**Comment by hgingras** [comment]
Hello, I tried to run with my environment with python 3.8.10 and I had the same error message. I uploaded the .sif image with apptainer. 

Module load nextflow/22.10.8

Running: nextflow run main.nf -c nextflow.config  --fastq test_data/porec_test.concatemers.fastq  --ref test_data/porec_test.fasta  --phased_vcf test_data/porec_test.phased_variants.vcf.gz  --cutter NlaIII  --vcf test_data/porec_test.phased_variants.vcf.gz --pairs  --chromunity --paired_end --filter_pairs --paired_end_minimum_distance 100  --paired_end_maximum_distance 200  --out_dir outputPoreC-test  -profile apptainer

I get an error with the first step: POREC:getVersions : fastcat: command not found.

So I guess apptainer sif file is not located.

I would need some help to set-up config file. 

Here is what I have added to config file:

In params list I added:

process.container = '/home/helene/scratch/Ticket/wf-pore-C/wf-pore-C_apptainer/wf-pore-c/wf-pore-c_latest.sif'
apptainer.enabled = true

In profiles list I added:

    }
    apptainer {
        apptainer {
            enabled = true
            autoMounts = true
        }
    }

Is there something else I need to do?

Best regards,

Helene

**Comment by sarahjeeeze** [comment]
Hi, we don't support apptainer and haven't tested it but perhaps if you are using a local img specify it along with the label eg.
``` 
process {
    withLabel:wfporec {
        container = '/home/helene/scratch/Ticket/wf-pore-C/wf-pore-C_apptainer/wf-pore-c/wf-pore-c_latest.sif'
    }

**Comment by hgingras** [comment]
Thanks for your suggestions. I figured out what I need to run the pipeline. Thanks for everything!

**Comment by sarahjeeeze** [comment]
Thanks, closing as resolved

**Comment by adRn-s** [comment]
For the record, I just had a similar error (workflow-glue command not found, makeReport rule failing). It was solved after updating to nextflow 23.10 :)
